qemu/tests/blockdev_mirror_cancel_ready_job_with_ioerror.py

```diff
@@ -0,0 +1,116 @@
+from provider.blockdev_mirror_nowait import BlockdevMirrorNowaitTest
+from provider.job_utils import get_event_by_condition
+
+
+class BlockdevMirrorCancelReadyIOError(BlockdevMirrorNowaitTest):
+    """
+    Cancel a ready job when target image is in error
+    """
+
+    def __init__(self, test, params, env):
+        params['filter-node-name'] = params['filter_node_name']
+        super(BlockdevMirrorCancelReadyIOError, self).__init__(test, params, env)
+
+    def _blockdev_add_images(self, images):
+        for tag in images:
+            params = self.params.object_params(tag)
+            devices = self.main_vm.devices.images_define_by_params(tag,
+                                                                   params,
+                                                                   'disk')
+            devices.pop()
+            for dev in devices:
+                if self.main_vm.devices.get_by_qid(dev.get_qid()):
+                    continue
+                ret = self.main_vm.devices.simple_hotplug(dev,
+                                                          self.main_vm.monitor)
+                if not ret[1]:
+                    self.test.fail("Failed to hotplug '%s': %s."
+                                   % (dev, ret[0]))
+
+    def _create_images(self, images):
+        for tag in images:
+            disk = self.source_disk_define_by_params(self.params, tag)
+            disk.create(self.params)
+            self.trash.append(disk)
+
+    def create_source_images(self):
+        """create source images of data images"""
+        self._create_images(self.params.objects("source_images"))
+
+    def create_target_images(self):
+        """create target images of mirror images"""
+        self._create_images(self.params.objects("target_images"))
+
+    def add_source_images(self):
+        """blockdev-add source images: protocol and format nodes only"""
+        self.create_source_images()
+        self._blockdev_add_images(self.params.objects("source_images"))
+
+    def add_target_images(self):
+        """blockdev-add target images: protocol and format nodes only"""
+        self.create_target_images()
+        # Fixme if blkdebug driver is supported completely in avocado-vt
+        args = {'node-name': 'drive_target', 'driver': 'qcow2', 'file': {'driver': 'blkdebug',
+                'image': {'driver': 'file', 'filename': '/root/avocado/data/avocado-vt/target.qcow2'},
```

Comment:
others can be hard-coded here except the filename, which may be different when using a python venv(--venv in kar), so we may have to get it by storage.get_image_filename or disk.image_filename, so what about:

def add_target_image(self):
tgt = self.disk_define_by_params(self.params, self._target_images[0])  # dont use source_disk_define_by_params
tgt.create(self.params) 
self.trash.append(tgt)
args = {'node-name': self._target_nodes[0], 'driver': tgt.image_format, xxx, 'filename': tgt.image_filename, xxx}
xxx

Reply:
@zhencliu, have set the target_filename a variable, thanks, zhenchao.